### PR TITLE
NX_COMPLEX - Update nxdlTypes.xsd

### LINKED ISSUE #1056: NX_COMPLEX: a new NeXus data type for complex numbers
The use of a complex number data type ([NX_COMPLEX](https://github.com/nexusformat/definitions/issues/1066)) was described at the 2022 NIAC Meeting.

### PR DESCRIPTION
**Pull request by sanbrock**
added NX_CCOMPLEX, NX_PCOMPLEX. NX_COMPLEX, and NX_QUATERNION

* closes #1056

### CODE REVIEW

File: nxdlTypes.xsd [hunk 1]
```diff
@@ -519,6 +523,48 @@
...
+		</xs:annotation>
+		<xs:sequence>
+			<xs:element name="real" type="xs:double"/>
+			<xs:element name="imaginary" type="xs:double"/>
+		</xs:sequence>
+	</xs:complexType>
+	
+	<xs:complexType name="NX_PCOMPLEX">
+		<xs:annotation>
+			<xs:documentation>Compound type polar representaion of complex numbers (amplitude and phase) in NeXus</xs:documentation>
```
Comment (PeterC-DLS) [inline nxdlTypes.xsd:538]: Specify units for angle component

File: nxdlTypes.xsd [hunk 2]
```diff
@@ -519,6 +523,48 @@
...
+			nxdl:NX_PCOMPLEX
+			" />
+	</xs:simpleType>
+	
+	<xs:complexType name="NX_QUATERNION">
+		<xs:annotation>
+			<xs:documentation>Compound type representaion of quaternion numbers (scalar,i,j,k) in NeXus</xs:documentation>
+		</xs:annotation>
+		<xs:sequence>
+			<xs:element name="scalar" type="xs:double"/>
```
Comment (PeterC-DLS) [inline nxdlTypes.xsd:561]: I prefer `real` here as it mirrors the complex number definition and a scalar/vector interpretation of a quarternoin may be confusing things by introducing a geometric meaning.

**Comment by mkuehbach** [comment]
xs:documentation have a typo in all four data types: should be "representation" instead of "representaion"

File: nxdlTypes.xsd (hunk 2 above)
Comment (sanbrock) [reply nxdlTypes.xsd]: done

**Comment by prjemian** [comment]
@PeterC-DLS This is ready for review.

**Comment by prjemian** [comment]
View of published documentation:
Clipboard01

File: nxdlTypes.xsd [hunk 3]
```diff
@@ -518,6 +522,58 @@
...
+		<xs:restriction base="nxdl:list_of_doubles">
+			<xs:length value='2' />
+		</xs:restriction>
+	</xs:simpleType>
+
+	<xs:simpleType name="NX_PCOMPLEX">
+		<xs:annotation>
+			<xs:documentation>
+				Compound type polar representation of complex numbers (amplitude
+				and phase *in radian*) in NeXus.
```
Comment (PeterC-DLS) [inline nxdlTypes.xsd:546]: *in radians*

**Comment by sanbrock** [comment]
NX_COMPLEX types are now represented as LIST of 2 doubles (in case of quaternions, 4 doubles). Trying to describe them as a proper compound type with labels, like 'r' for the real part and 'i' for the imaginary, one should define a SEQUENCE instead of a list.  
This requires the creation and use of ComplexTypes instead of SimpleTypes for our 'primitiveTypes'. Unfortunately, complexTypes are not supported to specify the TYPE of an ATTRIBUTE which we need for defining the NXDL attribute 'type':    
[https://github.com/nexusformat/definitions/blob/1056-NX_COMPLEX/nxdl.xsd?#L764](https://github.com/nexusformat/definitions/blob/1056-NX_COMPLEX/nxdl.xsd?#L764 )

File: nxdlTypes.xsd (hunk 3 above)
Comment (sanbrock) [reply nxdlTypes.xsd]: done

File: nxdlTypes.xsd [hunk 4]
```diff
@@ -518,6 +522,58 @@
...
+		<xs:union memberTypes="
+			nxdl:NX_CCOMPLEX  
+			nxdl:NX_PCOMPLEX
+			" />
+	</xs:simpleType>
+	
+	<xs:simpleType name="NX_QUATERNION">
+		<xs:annotation>
+			<xs:documentation>
+			Compound type representation of quaternion numbers (real,i,j,k) in NeXus.
```
[the remark this answers is not in the document]
Comment (sanbrock) [reply nxdlTypes.xsd]: done

**Review by PeterC-DLS — APPROVED** [review]
LGTM

**Comment by prjemian** [comment]
@PeterC-DLS Thanks for the review!

@sanbrock Merge when you are ready.

**Pull request merged**